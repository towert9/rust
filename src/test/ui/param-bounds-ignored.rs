// Copyright 2014 The Rust Project Developers. See the COPYRIGHT
// file at the top-level directory of this distribution and at
// http://rust-lang.org/COPYRIGHT.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

// must-compile-successfully
#![allow(dead_code, non_camel_case_types)]

use std::rc::Rc;

type SVec<T: Send> = Vec<T>;
type VVec<'b, 'a: 'b> = Vec<&'a i32>;
type WVec<'b, T: 'b> = Vec<T>;
type W2Vec<'b, T> where T: 'b = Vec<T>;

fn foo<'a>(y: &'a i32) {
    // If the bounds above would matter, the code below would be rejected.
    let mut x : SVec<_> = Vec::new();
    x.push(Rc::new(42));

    let mut x : VVec<'static, 'a> = Vec::new();
    x.push(y);

    let mut x : WVec<'static, & 'a i32> = Vec::new();
    x.push(y);

    let mut x : W2Vec<'static, & 'a i32> = Vec::new();
    x.push(y);
}

fn bar1<'a, 'b>(
    x: &'a i32,
    y: &'b i32,
    f: for<'xa, 'xb: 'xa> fn(&'xa i32, &'xb i32) -> &'xa i32)
{
    // If the bound in f's type would matter, the call below would (have to)
    // be rejected.
    f(x, y);
}

fn bar2<'a, 'b, F: for<'xa, 'xb: 'xa> Fn(&'xa i32, &'xb i32) -> &'xa i32>(
    x: &'a i32,
    y: &'b i32,
    f: F)
{
    // If the bound in f's type would matter, the call below would (have to)
    // be rejected.
    f(x, y);
}

fn bar3<'a, 'b, F>(
    x: &'a i32,
    y: &'b i32,
    f: F)
    where F: for<'xa, 'xb: 'xa> Fn(&'xa i32, &'xb i32) -> &'xa i32
{
    // If the bound in f's type would matter, the call below would (have to)
    // be rejected.
    f(x, y);
}

fn bar4<'a, 'b, F>(
    x: &'a i32,
    y: &'b i32,
    f: F)
    where for<'xa, 'xb: 'xa> F: Fn(&'xa i32, &'xb i32) -> &'xa i32
{
    // If the bound in f's type would matter, the call below would (have to)
    // be rejected.
    f(x, y);
}

struct S1<F: for<'xa, 'xb: 'xa> Fn(&'xa i32, &'xb i32) -> &'xa i32>(F);
struct S2<F>(F) where F: for<'xa, 'xb: 'xa> Fn(&'xa i32, &'xb i32) -> &'xa i32;
struct S3<F>(F) where for<'xa, 'xb: 'xa> F: Fn(&'xa i32, &'xb i32) -> &'xa i32;

struct S_fnty(for<'xa, 'xb: 'xa> fn(&'xa i32, &'xb i32) -> &'xa i32);

type T1 = Box<for<'xa, 'xb: 'xa> Fn(&'xa i32, &'xb i32) -> &'xa i32>;

fn main() {
    let _ : Option<for<'xa, 'xb: 'xa> fn(&'xa i32, &'xb i32) -> &'xa i32> = None;
    let _ : Option<Box<for<'xa, 'xb: 'xa> Fn(&'xa i32, &'xb i32) -> &'xa i32>> = None;
}
