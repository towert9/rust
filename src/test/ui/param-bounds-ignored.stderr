warning: bounds on generic type parameters are ignored in type aliases
  --> $DIR/param-bounds-ignored.rs:17:14
   |
17 | type SVec<T: Send+Send> = Vec<T>;
   |              ^^^^ ^^^^
   |
note: lint level defined here
  --> $DIR/param-bounds-ignored.rs:13:9
   |
13 | #![warn(ignored_generic_bounds)]
   |         ^^^^^^^^^^^^^^^^^^^^^^

warning: bounds on generic lifetime parameters are ignored in type aliases
  --> $DIR/param-bounds-ignored.rs:19:19
   |
19 | type VVec<'b, 'a: 'b+'b> = Vec<&'a i32>;
   |                   ^^ ^^

warning: bounds on generic type parameters are ignored in type aliases
  --> $DIR/param-bounds-ignored.rs:21:18
   |
21 | type WVec<'b, T: 'b+'b> = Vec<T>;
   |                  ^^ ^^

warning: where clauses are ignored in type aliases
  --> $DIR/param-bounds-ignored.rs:23:25
   |
23 | type W2Vec<'b, T> where T: 'b, T: 'b = Vec<T>;
   |                         ^^^^^  ^^^^^

warning: bounds on generic lifetime parameters are ignored in higher-ranked function types (i.e., `for`)
  --> $DIR/param-bounds-ignored.rs:44:22
   |
44 |     f: for<'xa, 'xb: 'xa> fn(&'xa i32, &'xb i32) -> &'xa i32)
   |                      ^^^

warning: bounds on generic lifetime parameters are ignored in higher-ranked trait bounds (i.e., `for`)
  --> $DIR/param-bounds-ignored.rs:52:34
   |
52 | fn bar2<'a, 'b, F: for<'xa, 'xb: 'xa> Fn(&'xa i32, &'xb i32) -> &'xa i32>(
   |                                  ^^^

warning: bounds on generic lifetime parameters are ignored in higher-ranked trait bounds (i.e., `for`)
  --> $DIR/param-bounds-ignored.rs:67:28
   |
67 |     where F: for<'xa, 'xb: 'xa> Fn(&'xa i32, &'xb i32) -> &'xa i32
   |                            ^^^

warning: bounds on generic lifetime parameters are ignored in higher-ranked trait bounds (i.e., `for`)
  --> $DIR/param-bounds-ignored.rs:79:25
   |
79 |     where for<'xa, 'xb: 'xa> F: Fn(&'xa i32, &'xb i32) -> &'xa i32
   |                         ^^^

warning: bounds on generic lifetime parameters are ignored in higher-ranked trait bounds (i.e., `for`)
  --> $DIR/param-bounds-ignored.rs:87:28
   |
87 | struct S1<F: for<'xa, 'xb: 'xa> Fn(&'xa i32, &'xb i32) -> &'xa i32>(F);
   |                            ^^^

warning: bounds on generic lifetime parameters are ignored in higher-ranked trait bounds (i.e., `for`)
  --> $DIR/param-bounds-ignored.rs:89:40
   |
89 | struct S2<F>(F) where F: for<'xa, 'xb: 'xa> Fn(&'xa i32, &'xb i32) -> &'xa i32;
   |                                        ^^^

warning: bounds on generic lifetime parameters are ignored in higher-ranked trait bounds (i.e., `for`)
  --> $DIR/param-bounds-ignored.rs:91:37
   |
91 | struct S3<F>(F) where for<'xa, 'xb: 'xa> F: Fn(&'xa i32, &'xb i32) -> &'xa i32;
   |                                     ^^^

warning: bounds on generic lifetime parameters are ignored in higher-ranked function types (i.e., `for`)
  --> $DIR/param-bounds-ignored.rs:94:29
   |
94 | struct S_fnty(for<'xa, 'xb: 'xa> fn(&'xa i32, &'xb i32) -> &'xa i32);
   |                             ^^^

warning: bounds on generic lifetime parameters are ignored in higher-ranked trait bounds (i.e., `for`)
  --> $DIR/param-bounds-ignored.rs:97:29
   |
97 | type T1 = Box<for<'xa, 'xb: 'xa> Fn(&'xa i32, &'xb i32) -> &'xa i32>;
   |                             ^^^

warning: bounds on generic lifetime parameters are ignored in higher-ranked function types (i.e., `for`)
   --> $DIR/param-bounds-ignored.rs:101:34
    |
101 |     let _ : Option<for<'xa, 'xb: 'xa> fn(&'xa i32, &'xb i32) -> &'xa i32> = None;
    |                                  ^^^

warning: bounds on generic lifetime parameters are ignored in higher-ranked trait bounds (i.e., `for`)
   --> $DIR/param-bounds-ignored.rs:103:38
    |
103 |     let _ : Option<Box<for<'xa, 'xb: 'xa> Fn(&'xa i32, &'xb i32) -> &'xa i32>> = None;
    |                                      ^^^

